kolena/_experimental/trace/trace.py

```diff
@@ -0,0 +1,239 @@
+# Copyright 2021-2024 Kolena Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import atexit
+import dataclasses
+import inspect
+import threading
+import time
+import uuid
+from collections import defaultdict
+from datetime import datetime
+from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Set
+from typing import Tuple
+
+import numpy as np
+import pandas as pd
+
+import kolena
+from kolena._utils.datatypes import DataObject
+from kolena.dataset import upload_results
+from kolena.dataset.dataset import _load_dataset_metadata
+from kolena.dataset.dataset import _upload_dataset
+from kolena.errors import NotFoundError
+
+THIRTY_SECONDS = 30
+KOLENA_DEFAULT_ID = "_kolena_id"
+KOLENA_TIMESTAMP_KEY = "_kolena_timestamp"
+KOLENA_TIME_ELAPSED_KEY = "_kolena_time_elapsed"
+
+
+def _serialize_item(value: Any):
+    if isinstance(value, (bool, str, int, float)) or value is None:
+        return value
+    elif isinstance(value, np.generic):  # numpy scalars are common enough to be worth specific handling
+        for base_type, numpy_type in [(bool, np.bool_), (int, np.integer), (float, np.inexact)]:
+            if isinstance(value, numpy_type):  # cast if there is a match, otherwise fallthrough
+                return base_type(value)
+    elif isinstance(value, Dict):
+        return {key: _serialize_item(subvalue) for key, subvalue in value.items()}
+    elif isinstance(value, DataObject):
+        return value._to_dict()
+    elif dataclasses.is_dataclass(value):
+        return _serialize_item(dataclasses.asdict(value))
+    elif hasattr(value, "__dict__"):
+        return _serialize_item(value.__dict__)
+    elif isinstance(value, (List, Set, Tuple, pd.Series, np.ndarray)):
+        return [_serialize_item(item) for item in value]
+    else:
+        return str(value)
+
+
+class _Trace:
```

Comment:
why do we use class at here?

Reply:
We want to keep send in the upload datapoint and result request in batch to not overwhelm our server, temporarily storing request and responses in class